src/daily_output.ts

```diff
@@ -0,0 +1,474 @@
+		);
+		const hasReachedMinimumGroupEffort = groupEffortParticipants.length >= 50;
+		this.data.ranks.daily.hasReachedMinimumGroupEffort = hasReachedMinimumGroupEffort;
+		this.data.ranks.weekly.hasReachedMinimumGroupEffort = hasReachedMinimumGroupEffort;
+
+		// compute group effort rewards
+		if (hasReachedMinimumGroupEffort) {
+			/**
+			 * apply tiebreakers:
+			 * - by total upload volume
+			 * - by total tips sent
+			 * - by block since participating (i.e. has reached the minimum weekly data)
+			 */
+			const shuffledTies = groupEffortParticipants.sort((address_a, address_b) => {
+				const walletStat_a = this.data.wallets[address_a];
+				const walletStat_b = this.data.wallets[address_b];
+				const volumeDiff = walletStat_a.weekly.byteCount - walletStat_b.weekly.byteCount;
+				const tipsDiff = walletStat_a.weekly.tips - walletStat_b.weekly.tips;
+				const blockSinceParticipatingDiff =
+					// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+					walletStat_a.weekly.blockSinceParticipating! - walletStat_b.weekly.blockSinceParticipating!;
+				return volumeDiff || tipsDiff || blockSinceParticipatingDiff;
+			});
+
+			shuffledTies.forEach((address, index) => (this.data.wallets[address].daily.rankPosition = index + 1));
+
+			const top50 = shuffledTies.slice(0, 49);
+
+			this.data.ranks.daily.groupEffortRewards = top50.map((address, index) => {
+				return { address, rewards: GROUP_EFFORT_REWARDS[index] };
+			});
+		}
+
+		// compute streak rewards
+		// const stakedPSTHolders = Object.keys(this.data.PSTHolders);
+		// TODO: determine if the wallets has uploaded data for 7 days in a row
+
+		this.data.lastUpdated = Date.now();
+	}
+
+	private changeInPercentage(prev: number, curr: number): number {
+		if (prev === 0) {
+			if (curr === 0) {
+				// Both zero, there's no change
+				return 0;
+			} else {
+				// Previous is zero, current is greater: 100% change
+				return 1;
+			}
+		} else {
+			return (curr - prev) / prev;
+		}
+	}
+
+	/**
+	 * Will copy the data of today into yesterday's and clear the today's data
+	 */
+	private resetDay(): void {
+		for (const address in this.data.wallets) {
+			this.data.wallets[address].yesterday = this.data.wallets[address].daily;
+			this.data.wallets[address].daily = {
+				byteCount: 0,
+				changeInPercentage: 0,
+				fileCount: 0,
+				rankPosition: 0,
+				tokensEarned: 0,
+				tips: 0
+			};
+			this.data.ranks.daily = {
+				hasReachedMinimumGroupEffort: false,
+				groupEffortRewards: [],
+				streakRewards: []
+			};
+		}
+	}
+
+	/**
+	 * Will copy the data of the current week into the past one and clear the data of the current week
+	 */
+	private resetWeek(): void {
+		for (const address in this.data.wallets) {
+			this.data.wallets[address].lastWeek = this.data.wallets[address].weekly;
+			this.data.wallets[address].weekly = {
+				byteCount: 0,
+				changeInPercentage: 0,
+				fileCount: 0,
+				rankPosition: 0,
+				tokensEarned: 0,
+				tips: 0
+			};
```

Comment:
Lets store some of these repeated templates as a value somewhere, e.g:

```ts
const walletStatTemplate = {
				byteCount: 0,
				changeInPercentage: 0,
				fileCount: 0,
				rankPosition: 0,
				tokensEarned: 0,
				tips: 0
			}
```

Reply:
fa01df2ecefb5099b5a9e1e1e2fa1f5abf00cfe9